### PR TITLE
DDF-2847 Adds unsaved workspace indicators to the UI

### PR DESCRIPTION
**Pull request by andrewkfiedler**
#### What does this PR do?
 - Updates various views to listen to workspace changes and update to show an indication of unsaved workspace changes.
 - The css gets a little trickier than it appears it needs to be, mostly to support the same look and feel on IE10, IE11, and IE Edge.  These browsers do not support transitions on height or width that involve calc.  Unfortunately, that means we have to get clever.
 - Updates focus styling to be more consistent across browsers.  IE does not support outline-offset.
 - Breaks out part of workspace-item into workspace-details.  Updates workspace-item html to be simpler.

#### Who is reviewing it (please choose AT LEAST two reviewers that need to approve the PR before it can get merged)?
@pklinef 
@clockard 
@djblue 
@rzwiefel 
@andrewkfiedler

#### How should this be tested? (List steps with links to updated documentation)
Do this both logged in, then while logged out.  
Make a workspace.  Verify that no indicator appears.  Make a change to the workspace (change the title, add a query, remove a query, change a query, add or remove a bookmark).  Verify the indicator appears for these actions.  Run a query.  Verify the indicator does not appear when a query is run.

Try saving from the workspace route, then from the workspaces route.  Verify that deleting the only unsaved workspace removes indicators.  

#### What are the relevant tickets?
https://codice.atlassian.net/browse/DDF-2847

#### Screenshots (if appropriate)
screen shot 2017-03-06 at 12 11 06 am
screen shot 2017-03-06 at 12 10 55 am
screen shot 2017-03-06 at 12 10 44 am

### CODE REVIEW

**Comment by andrewkfiedler** [comment]
Verified and tested on all supported browsers:  IE10, IE11, Edge, Chrome, Firefox ESR, and Firefox.

File: catalog/ui/catalog-ui-search/src/main/webapp/component/content-title/content-title.hbs [hunk 1]
```diff
@@ -11,5 +11,7 @@
  *
  **/
  --}}
-<input placeholder="Workspace Title" value="{{currentWorkspace.title}}">
-<div class="title-display">{{currentWorkspace.title}}</div>
\ No newline at end of file
+<input placeholder="Workspace Title" value="{{currentWorkspace.title}}" data-help="This is the title of the workspace you are currently in.
+If you have permission, you can click here to start editing the title.">
+<pre class="title-display">{{currentWorkspace.title}}</pre>
+<span class="title-saved">*</span>
```
Comment (rzwiefel) [inline catalog/ui/catalog-ui-search/src/main/webapp/component/content-title/content-title.hbs]: @andrewkfiedler is it possible to add the contextual help to the `*` span ?

**Comment by brendan-hofmann** [comment]
Will this update the title text (in the browser tab) as well?

File: catalog/ui/catalog-ui-search/src/main/webapp/js/model/Workspace.js [hunk 2]
```diff
@@ -213,30 +238,49 @@ define([
...
-                    })
+                    .reduce(function (blob, workspace) {
+                        blob[workspace.get('id')] = workspace.toJSON();
+                        return blob;
+                    }, {})
                     .value();
 
                 window.localStorage.setItem('workspaces', JSON.stringify(localWorkspaces));
             },
+            convertOldFormat: function(localWorkspaceJSON){
```
Comment (pklinef) [inline catalog/ui/catalog-ui-search/src/main/webapp/js/model/Workspace.js]: We should consider adding a version to know what need to be converted.  This is going to be very confusing if we continue to change how workspaces are stored in the future.

File: catalog/ui/catalog-ui-search/src/main/webapp/component/workspace-item/workspace-item.view.js [hunk 3]
```diff
@@ -71,10 +63,12 @@ define([
                 }
             });
         },
-        serializeData: function() {
-            var workspacesJSON = this.model.toJSON();
-            workspacesJSON.niceDate = moment(workspacesJSON.modified).fromNow();
-            return workspacesJSON;
+        handleSaved: function(){
+            this.$el.toggleClass('is-saved', this.model.isSaved());
+            this.$el.find('.choice-save').attr('tabindex', this.model.isSaved() ? -1 : null);
```
Comment (andrewkfiedler) [inline catalog/ui/catalog-ui-search/src/main/webapp/component/workspace-item/workspace-item.view.js]: For animations, you can't do a full rerender.

File: catalog/ui/catalog-ui-search/src/main/webapp/component/workspace-menu/workspace-menu.less [hunk 4]
```diff
@@ -5,23 +5,90 @@
...
+    overflow: hidden;
+    width: @minimumButtonSize;
+  }
+
+  .content-save .fa {
+    vertical-align: middle;
+  }
+
+  .content-save .fa-floppy-o {
+    transition: opacity 2*@coreTransitionTime ease-out 0*@coreTransitionTime,
```
Comment (andrewkfiedler) [inline catalog/ui/catalog-ui-search/src/main/webapp/component/workspace-menu/workspace-menu.less]: We might be able to, a lot of this was rushed to get it into the backport.

File: catalog/ui/catalog-ui-search/src/main/webapp/js/model/Workspace.js (hunk 2 above)
Comment (andrewkfiedler) [inline catalog/ui/catalog-ui-search/src/main/webapp/js/model/Workspace.js]: Not thinking we'll be changing from the current method of storing them, but we can name it something like convertArrayStoredWorkspacesToMapStoredWorkspaces.

File: catalog/ui/catalog-ui-search/src/main/webapp/component/workspace-item/workspace-item.less [hunk 5]
```diff
@@ -1,141 +1,112 @@
-@{customElementNamespace}workspace-item.as-grid {
+@{customElementNamespace}workspace-item {
   .customElement;
+  overflow: hidden;
   display: inline-block;
-  width: auto;
+  width: 18rem;
```
Comment (andrewkfiedler) [inline catalog/ui/catalog-ui-search/src/main/webapp/component/workspace-item/workspace-item.less]: It looks fine.  Haven't had a chance to go back and update some of the px based margins, but they work well under zoom as is.

File: catalog/ui/catalog-ui-search/src/main/webapp/component/content-title/content-title.hbs [hunk 6]
```diff
@@ -11,5 +11,7 @@
  *
  **/
  --}}
-<input placeholder="Workspace Title" value="{{currentWorkspace.title}}">
-<div class="title-display">{{currentWorkspace.title}}</div>
\ No newline at end of file
+<input placeholder="Workspace Title" value="{{bind selector="> input" event="change:currentWorkspace" key=(path 'currentWorkspace' 'title')}}" data-help="This is the title of the workspace you are currently in.
+If you have permission, you can click here to start editing the title.">
+<pre class="title-display">{{bind selector="> .title-display" event="change:currentWorkspace" key=(path 'currentWorkspace' 'title')}}</pre>
```
Comment (rzwiefel) [inline catalog/ui/catalog-ui-search/src/main/webapp/component/content-title/content-title.hbs]: Wow you can path to nested elements? Great job here @andrewkfiedler ! 😉

File: catalog/ui/catalog-ui-search/src/main/webapp/component/navigation-left/navigation-left.less [hunk 7]
```diff
@@ -1,10 +1,18 @@
...
     button {
         text-align: center;
         width: @minimumButtonSize;
         height: 100%;
     }
+
+    button > .navigation-indicator {
+        position: absolute;
+        right: 0px;
+        top: -5px;
```
Comment (rzwiefel) [inline catalog/ui/catalog-ui-search/src/main/webapp/component/navigation-left/navigation-left.less]: Should this be sized in rem so when you resize page the indicator stays where it is relative to the text its next to?

**Pull request merged**